_posts/2020-09-27-spring-boot-12-factor-app.md

```diff
@@ -0,0 +1,169 @@
+
+## Dependencies
+> Explicitly declare and isolate dependencies
+
+The most likely dependencies of an application are the libraries pulled from open source repositories or built inhouse by other teams. Dependencies could also take the form of specific software installed in the host system. The dependencies are declared in external files leveraging dependency management tools of the platform.
+
+For the Spring Boot application, we declare the library dependencies in `pom.xml` or Gradle depending on whether we are using Maven or Gradle. Our Spring Boot application uses `spring-boot-starter-web` as one of its dependencies. The declarative form of this dependency is added to the `pom.xml`: 
+
+```xml
+    <dependency>
+      <groupId>org.springframework.boot</groupId>
+      <artifactId>spring-boot-starter-web</artifactId>
+    </dependency>
+```
+
+This principle is an evolution from an earlier practice of sharing libraries across applications by storing them in a shared classpath. But using this approach introduced a coupling with the configuration of the host system. 
+
+The declarative style of specifying dependencies removed this coupling. In the context of using Spring Boot, when using a dependency tool like Maven/Gradle also allows us in :
+ - ***Versioning*** by declaring specific versions of the dependencies with which our application works, and
+ - ***Isolating*** them by bundling them with the application.
+
+## Config - Externalizing Configuration Properties
+> Store config in the environment
+
+Few examples of configuration data are database connection URL and credentials, and URLs of services on which an application depends. These most often have different values across environments. If these are hard-coded in the code or property files bundled with the application, we need to update the application for deploying to different environments. 
+
+Instead, a better approach is to [externalizing](https://reflectoring.io/externalize-configuration/https://reflectoring.io/externalize-configuration/) the configuration using environment variables. The values of the environment variables are provided at runtime. We can provide the values from the command line if the application is run standalone. The default behavior in Spring Boot applications is to apply the values from environment variables to override any values declared in property files.
+
+## Backing Services - Pluggable Datasources, and Queues
+> Treat backing services as attached resources
+
+Backing services should be attached and replaceable instead of embedded in the code. The use of specifications like JPA helps us achieve this for RDBMS databases. But in the absence of specifications, some code creeps into the code although we can keep them separate with abstraction layers. Similar to JPA, we can use JMS for messaging and SMTP for mails.
```

Comment:
I don't understand this paragraph, please rephrase. 

I think the "Pluggability" is best explained with
* using an Abstraction like JPA over a data source like an RDBS
* using configuration properties (like a JDBC URL) to configure the connection

This way, we can just change the JDBC URL to swap out the database. And we can swap out the underlying database by changing the dependency.

Reply:
Also, why should backing services be attached and replacable?